source/transfer_list.rst

```diff
@@ -146,15 +146,49 @@ The TE header is defined in :numref:`tab_te_header`.
      - The exact size of the data content in bytes, not including inter-TE padding. May be 0.
 
 
-Entry type ranges
------------------
-
-The content of the data section is determined by the tag id. The tag id space contains two ranges:
-
- #. Standard range, and
- #. Non-standard range
-
-The `tag_id` ranges are described in :numref:`tab_tag_id_ranges`.
+Entry type allocation
+---------------------
+
+Tag IDs must be allocated in this specification before use. A new tag ID can be
+allocated by submitting a pull request to this repository that adds a
+description of the respective TE data layout to this specification. Tag IDs do
+not have to be allocated in order. Submitters are encouraged to try to group
+tag IDs together in logical clusters at 256-aligned boundaries (e.g. all tags
```

Comment:
256 seems like a lot. Perhaps this can be handled by adding a comment above the allocation which indicates how many are likely to be needed? For many projects, 16 might be enough.

I suggest "16-aligned (or if necessary 256-aligned) boundaries"

Reply:
There are 32768 groups of 256 tags in the standardized range (and almost again that many if we start using the reserved range). Do you really think it's likely that we'll find that many independent stakeholders who want to add their own things here?

I really think 256 is the most convenient trade-off here -- it's large enough that few projects will likely need to allocate a second one, but still gives us so many independent blocks that I think it's extremely unlikely we'll run out (and, after all, in the extremely unlikely event we do run out we can still start reusing the high 128 bits of old blocks with just 3 entries where nothing new has been added in years... this is just guidance, not something that is forever set in stone). If anything I'd wonder if it should go up to 512 or 1024, but in the end it's not that much of a difference.

But I think 16 is so small that it will likely require most projects to allocate multiple fields and kinda defeat the "group similar stuff together" idea. I really don't think we need to be that overly frugal with this. 23 bits is a _lot_ of ID space and there's no prize waiting for us if 10 years from now 99% of it is still unused.

Reply:
I agree that this is mostly guidance and we can adapt as we go.
It's also true that 32768 groups of 256 tags is a lot. We'll realistically never run out of these groups.

Reply:
Sorry, I really don't understand the second part. Doesn't 0x101, 0x0102, 0x103 vs. 0x0201, 0x0202, 0x0203 make it much easier to see which tags are grouped together (0x1 vs. 0x2 is the group prefix and the last two hex digits are within the group)? In what way is that "harder" than being able to tell that 31 is in a different group than 32? I mean, most of us probably know the multiples of 16 pretty well but it's still easier to not have to do the math at all because the split is already built into the digits?

Reply:
Well I believe in practice we would end up with:

100, 101, 102
200
300, 301, 302, 303, 304, 305, 306, 307
400

etc.

which is using up a lot of space and making the numbers 10x larger than they need to be, just in case a project generates 100 (256 decimal) tags

I would rather have things closer together. I suppose we could allow projects to take 16 groups if they want to, then we get the best of both worlds

Reply:
Please see the calculations about how many groups of 256 tags can fit into the available space -- are you really concerned that this specification will draw tens of thousands of different firmware projects and SoC vendors? I mean, I am really hoping this will take off too, but that number still doesn't seem at all realistic to me. We don't gain anything from leaving all this ID space unused "just in case". I do however think that one consumer needing more than 16 tags is going to be pretty common (e.g. the handoff list from coreboot to its payload has 50+ tags at the moment).

> I would rather have things closer together. I suppose we could allow projects to take 16 groups if they want to, then we get the best of both worlds

If we should actually run out of groups (which I think is astronomically unlikely with 32K), we can still start reusing space from those that haven't been filled up much. Remember, this is meant to be a guideline, not a hard lease. We're not giving those 256 tags away to someone else who will never give them back, we're just saying "please pick a start number on this alignment and try to group all your stuff together there for as long as it fits".

I don't understand what the benefit of "having things closer together" is supposed to be. The whole point is that all groups are independent and there is no order or relationship between different groups.

Reply:
I am not worried about the amount of space. I'm just trying to avoid spreading out a lot over that space.

Another option would be to just allocate things sequentially and not worry about having each project's numbers next to each other. Would that work?

Reply:
> I'm just trying to avoid spreading out a lot over that space.

I really don't understand... _why?_ What's the problem with spreading out? It's an ID space! It doesn't matter where in there you allocate IDs, other than for aesthetics and to more easily recognize patterns when reading raw IDs in a hex dump.

> Another option would be to just allocate things sequentially and not worry about having each project's numbers next to each other. Would that work?

Yes, it would, it would just look a lot more messy. I really don't understand why you're preferring this solution. I think grouping has a clear aesthetics/recognizability benefit, and it may also reduce collision churn when many new ID PRs are submitted and reviewed at once. What's the benefit of not "speading out a lot over that space" that outweighs these?

Reply:
Again, I don't feel strongly here. FWIW I also don't understand the issue with spreading out the tag ID space. The compiler efficiency argument doesn't seem strong to me (assuming it's valid at all).

Reply:
I am OK with clumping them together, but I think clumps of 16 is enough. The compiler argument is valid on 32-bit ARM at least, where small numbers can be set in a register. But in general, smaller numbers are easier to read for humans. For example, 46 is easier that 106 just from the smaller number of digits. One of the things I hate about UUIDs is that they are so long. So why make things longer than necessary, just to keep the computer happy? These tags are for humans and we will be seeing them in binary dumps for years.

So that is the 'why'.

Most likely we'll have 5-6 projects using this initially, with fewer than 100 tags in use. Keeping these somewhat together seems best to me.

Reply:
If I'm to give my opinion, we're discussing a guidance statement that we'll have the liberty to adapt as we go along.

I see the point of both 16 and 256 groupings.
It is best to have tag ids that are more human-readable. On the other hand having 256 groupings does give more room for each "cluster" to grow.
Is there a compromise value? Say 64 perhaps?

I don't think that this discussion should gate us from a 0.9 release, and starting to work on the implementations.

Reply:
> So why make things longer than necessary, just to keep the computer happy? These tags are for humans and we will be seeing them in binary dumps for years.

It's usability for humans that I'm trying to argue for here! For the computer this is all the same. I find it very contrived to argue that there was any difference in readability between 46 and 106, I think for anything under 5 digits it really doesn't make a difference to human readers. But it does make a notable difference whether you can say "oh look, this is a 0x03xx ID, that means it must be one of the U-Boot tags" or you say "gee, I can't remember, was 0x024x one of the 3 coreboot prefixes or maybe one of the 4 U-Boot ones...". I don't think it's fair to compare an (at _most_, if this spec is far more successful than we could hope for) 6-digit hex number to a UUID in terms of human readability.

> Is there a compromise value? Say 64 perhaps?

I think that would be losing the readability advantage of breaking the ID at a hex boundary.

Reply:
That's all fine as a differing opinion, but I have answered your questions with a reasonable response.

Going back to my original point, how many are likely to be needed? For many projects, 16 might be enough? You didn't address the efficiency angle. Also I am a reader and I prefer shorted numbers to longer ones.

If we cannot agree this here, then I propose dropping all mention of how they will be allocated. 

We can also do it on the fly...e.g. if coreboot needs a range of 0x100 tags, it can have its own area allocated. For projects that don't, they can use a smaller area of 0x10.

Reply:
I'm not keen on dropping all mention of how they will be allocated but I agree we don't necessarily need a single block size. Could we just say to try to group in clusters of 16 or 256 aligned boundaries, depending on the likely number of TEs required?

Reply:
Yes that sounds good to me, using 16 or 256 depending on likely number ot TEs.

Reply:
Updated the text although I am concerned about the logistics of this (e.g. if one project starts grabbing tags at 0x200, 0x201, 0x202... how will the next project know whether they should continue at 0x210 or 0x300?). Guess we can argue about that in the individual submission reviews later.